### ISSUE TITLE
Autodiff panic during training, when run on multiple GPUs

### ISSUE THREAD

**Issue by marcusbuffett** [tracel-ai/burn#3513]
**Describe the bug**
When running on a single GPU, Burn does fine. But when I run on multiple GPUs, I get the following error, after some time has passed during training:

```
A fatal error happened, you can check the experiment logs here => 'checkpoints/experiment.log'
=============
                                                                                                                                   
thread '<unnamed>' panicked at /home/ubuntu/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/runtime/server.rs:34:48:
Node should have a step registered, did you forget to call `Tensor::register_grad` on the tensor where you need gradients?
stack backtrace:
   0:     0x5a291f4e0772 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hf435e8e9347709a8
   1:     0x5a291f50b1d3 - core::fmt::write::h0a51fad3804c5e7c
   2:     0x5a291f4dbf43 - std::io::Write::write_fmt::h9759e4151bf4a45e
   3:     0x5a291f4e05c2 - std::sys::backtrace::BacktraceLock::print::h1ec5ce5bb8ee285e
   4:     0x5a291f4e1d9c - std::panicking::default_hook::{{closure}}::h5ffefe997a3c75e4
   5:     0x5a291f4e1b9f - std::panicking::default_hook::h820c77ba0601d6bb
   6:     0x5a291f2adbc2 - <burn_train::learner::application_logger::FileApplicationLoggerInstaller as burn_train::learner::application_logger::ApplicationLoggerInstaller>::install::{{closure}}::hde491d3f74c5abc6
   7:     0x5a291f2abd18 - burn_train::renderer::tui::renderer::TuiMetricsRenderer::new::{{closure}}::h7be87050786bbb1e
   8:     0x5a291f4e27fb - std::panicking::rust_panic_with_hook::h8b29cbe181d50030
   9:     0x5a291f4e24ea - std::panicking::begin_panic_handler::{{closure}}::h9f5b6f6dc6fde83e
  10:     0x5a291f4e0c79 - std::sys::backtrace::__rust_end_short_backtrace::hd7b0c344383b0b61
  11:     0x5a291f4e217d - __rustc[5224e6b81cd82a8f]::rust_begin_unwind
  12:     0x5a291e449ef0 - core::panicking::panic_fmt::hc49fc28484033487
  13:     0x5a291e449ecb - core::option::expect_failed::hfe7afbd436ce9c45
  14:     0x5a291f27e40e - burn_autodiff::runtime::server::AutodiffServer::backward::h292a4feab65c2c7a
  15:     0x5a291ebe599c - <burn_autodiff::runtime::mutex::MutexClient as burn_autodiff::runtime::client::AutodiffClient>::backward::h891ab18fa559a4c0
  16:     0x5a291ee7ad3e - burn_tensor::tensor::api::autodiff::<impl burn_tensor::tensor::api::base::Tensor<B,_>>::backward::h2d9e61ac337d8980
  17:     0x5a291ed34eec - <oxi::model::OXIModel<B> as burn_train::learner::train_val::TrainStep<oxi::dataset::ChessBatch<B>,oxi::chess_output::ChessOutput<B>>>::step::he741a7a0c8a5886c
  18:     0x5a291ec21c37 - std::sys::backtrace::__rust_begin_short_backtrace::h0a808765edc33abd
  19:     0x5a291ebd72eb - core::ops::function::FnOnce::call_once{{vtable.shim}}::h631027a8241080f3
  20:     0x5a291f4e5b3f - std::sys::pal::unix::thread::Thread::new::thread_start::h1ff51d6e85162efd
  21:     0x7261d7294ac3 - start_thread
                               at ./nptl/pthread_create.c:442:8
  22:     0x7261d7326850 - __GI___clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81:0
  23:                0x0 - <unknown>
```

This is the experiment.log:

```
2025-08-12T00:53:12.031188Z  INFO burn_train::learner::train_val: Fitting the model:
 OXIModel {
  board_embed: Linear {d_input: 125, d_output: 384, bias: true, params: 48384}
  pos_embed: Embedding {n_embedding: 64, d_model: 384, params: 24576}
  global_embed: Linear {d_input: 7, d_output: 384, bias: true, params: 3072}
  blocks: Vec<0..0>
  norm: LayerNorm {d_model: 384, epsilon: 0.00001, params: 768}
  policy_head: Linear {d_input: 384, d_output: 76, bias: true, params: 29260}
  value_head: Linear {d_input: 384, d_output: 3, bias: true, params: 1155}
  side_info_head: Linear {d_input: 384, d_output: 13, bias: true, params: 5005}
  time_usage_head: Linear {d_input: 384, d_output: 6, bias: true, params: 2310}
  side_info_bce: BinaryCrossEntropyLoss {weights: None, smoothing: None, logits: true}
  policy_uncertainty: ParamTensor {rank: 1, shape: [1], kind: float}
  value_uncertainty: ParamTensor {rank: 1, shape: [1], kind: float}
  side_info_uncertainty: ParamTensor {rank: 1, shape: [1], kind: float}
  time_usage_uncertainty: ParamTensor {rank: 1, shape: [1], kind: float}
  params: 114534
}
2025-08-12T00:53:12.163929Z  INFO burn_train::learner::epoch: Executing training step for epoch 1 on devices [Cuda(0), Cuda(1), Cuda(2), Cuda(3), Cuda(4), Cuda(5), Cuda(6), Cuda(7)]
2025-08-12T00:53:14.500919Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:14.500984Z  INFO cubecl_runtime::tune::tune_cache: Loaded 76 autotune cached entries
2025-08-12T00:53:15.969603Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:15.969660Z  INFO cubecl_runtime::tune::tune_cache: Loaded 29 autotune cached entries
2025-08-12T00:53:15.970553Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:15.970606Z  INFO cubecl_runtime::tune::tune_cache: Loaded 29 autotune cached entries
2025-08-12T00:53:15.972939Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:15.972997Z  INFO cubecl_runtime::tune::tune_cache: Loaded 67 autotune cached entries
2025-08-12T00:53:15.973899Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:15.973948Z  INFO cubecl_runtime::tune::tune_cache: Loaded 29 autotune cached entries
2025-08-12T00:53:16.347358Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:16.347410Z  INFO cubecl_runtime::tune::tune_cache: Loaded 89 autotune cached entries
2025-08-12T00:53:16.392585Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:16.392632Z  INFO cubecl_runtime::tune::tune_cache: Loaded 29 autotune cached entries
2025-08-12T00:53:16.687563Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:16.687610Z  INFO cubecl_runtime::tune::tune_cache: Loaded 29 autotune cached entries
2025-08-12T00:53:16.688348Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:16.688391Z  INFO cubecl_runtime::tune::tune_cache: Loaded 29 autotune cached entries
2025-08-12T00:53:17.341681Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:17.341725Z  INFO cubecl_runtime::tune::tune_cache: Loaded 31 autotune cached entries
2025-08-12T00:53:17.375175Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:17.375220Z  INFO cubecl_runtime::tune::tune_cache: Loaded 56 autotune cached entries
2025-08-12T00:53:17.388890Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:17.388931Z  INFO cubecl_runtime::tune::tune_cache: Loaded 33 autotune cached entries
2025-08-12T00:53:17.861596Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:17.861647Z  INFO cubecl_runtime::tune::tune_cache: Loaded 80 autotune cached entries
2025-08-12T00:53:17.872330Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:17.872366Z  INFO cubecl_runtime::tune::tune_cache: Loaded 31 autotune cached entries
2025-08-12T00:53:17.895801Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:17.895842Z  INFO cubecl_runtime::tune::tune_cache: Loaded 31 autotune cached entries
2025-08-12T00:53:18.161394Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:18.161436Z  INFO cubecl_runtime::tune::tune_cache: Loaded 32 autotune cached entries
2025-08-12T00:53:18.182506Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:18.182556Z  INFO cubecl_runtime::tune::tune_cache: Loaded 18 autotune cached entries
2025-08-12T00:53:18.264324Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:18.264360Z  INFO cubecl_runtime::tune::tune_cache: Loaded 18 autotune cached entries
2025-08-12T00:53:18.394980Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:18.395018Z  INFO cubecl_runtime::tune::tune_cache: Loaded 28 autotune cached entries
2025-08-12T00:53:18.617938Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:18.617984Z  INFO cubecl_runtime::tune::tune_cache: Loaded 38 autotune cached entries
2025-08-12T00:53:18.686485Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:18.686537Z  INFO cubecl_runtime::tune::tune_cache: Loaded 18 autotune cached entries
2025-08-12T00:53:18.758690Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:18.758736Z  INFO cubecl_runtime::tune::tune_cache: Loaded 18 autotune cached entries
2025-08-12T00:53:19.071762Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:19.071801Z  INFO cubecl_runtime::tune::tune_cache: Loaded 18 autotune cached entries
2025-08-12T00:53:19.321365Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:19.321403Z  INFO cubecl_runtime::tune::tune_cache: Loaded 18 autotune cached entries
2025-08-12T00:53:20.490704Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:20.490746Z  INFO cubecl_runtime::tune::tune_cache: Loaded 31 autotune cached entries
2025-08-12T00:53:21.245105Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:21.245141Z  INFO cubecl_runtime::tune::tune_cache: Loaded 3 autotune cached entries
2025-08-12T00:53:21.318567Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:21.318610Z  INFO cubecl_runtime::tune::tune_cache: Loaded 3 autotune cached entries
2025-08-12T00:53:21.366405Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:21.366436Z  INFO cubecl_runtime::tune::tune_cache: Loaded 7 autotune cached entries
2025-08-12T00:53:21.464800Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:21.464834Z  INFO cubecl_runtime::tune::tune_cache: Loaded 3 autotune cached entries
2025-08-12T00:53:21.584242Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:21.584280Z  INFO cubecl_runtime::tune::tune_cache: Loaded 3 autotune cached entries
2025-08-12T00:53:21.720763Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:21.720798Z  INFO cubecl_runtime::tune::tune_cache: Loaded 3 autotune cached entries
2025-08-12T00:53:21.967063Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:21.967097Z  INFO cubecl_runtime::tune::tune_cache: Loaded 3 autotune cached entries
2025-08-12T00:53:25.619376Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:25.619420Z  INFO cubecl_runtime::tune::tune_cache: Loaded 13 autotune cached entries
2025-08-12T00:53:32.805236Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:32.805276Z  INFO cubecl_runtime::tune::tune_cache: Loaded 5 autotune cached entries
2025-08-12T00:53:38.601216Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:38.601259Z  INFO cubecl_runtime::tune::tune_cache: Loaded 5 autotune cached entries
2025-08-12T00:53:44.516253Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:44.516297Z  INFO cubecl_runtime::tune::tune_cache: Loaded 5 autotune cached entries
2025-08-12T00:53:50.399855Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:50.399895Z  INFO cubecl_runtime::tune::tune_cache: Loaded 5 autotune cached entries
2025-08-12T00:53:56.246721Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:53:56.246761Z  INFO cubecl_runtime::tune::tune_cache: Loaded 5 autotune cached entries
2025-08-12T00:54:02.176647Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:54:02.176692Z  INFO cubecl_runtime::tune::tune_cache: Loaded 5 autotune cached entries
2025-08-12T00:54:08.164347Z  INFO cubecl_runtime::tune::tune_cache: Load autotune cache ...
2025-08-12T00:54:08.164388Z  INFO cubecl_runtime::tune::tune_cache: Loaded 11 autotune cached entries
2025-08-12T00:55:44.219474Z ERROR burn_train::learner::application_logger: PANIC => panicked at /home/ubuntu/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/runtime/server.rs:34:48:
Node should have a step registered, did you forget to call `Tensor::register_grad` on the tensor where you need gradients?

```

**To Reproduce**
My project is open-source here: https://github.com/marcusbuffett/oxi

It should be as simple as running `cargo run -- download-all` (which you don't need to wait for) followed by `cargo run --release -- --max-samples=200000 num-devices=8` (assuming you have 8 devices available), on a linux machine with Cuda GPUs. It'll crash after a few minutes

**Desktop (please complete the following information):**
OS: Ubuntu
GPUs: 8xH100

**Comment by marcusbuffett** [comment]
Tried removing things until virtually nothing of my model was left, but no dice. I think the issue is in [rope.rs](https://github.com/marcusbuffett/oxi/blob/main/src/rope.rs), since I didn't see the error when running with `--disable-rope`, but I don't see anything suspicious in there, and it works fine with one GPU (or even multiple GPUs for a few thousand iterations)

**Comment by jonboh** [comment]
I'm reproducing this same panic on LibTorch (backend) on CPU on the test suite of the project I'm working on.
`Node should have a step registered, did you forget to call Tensor::register_grad on the tensor where you need gradients?`
Unfortunately I don't yet have a deterministic way to reproduce it, right now, running my training test suite continously ends up generating this panic (most of the time the test training finish correctly).
I'll test with https://github.com/tracel-ai/burn/pull/3793, and report if it fixes the issue.

**Comment by laggui** [comment]
@jonboh the fix should be included in 0.19.0 which was just released yesterday.

Please let me know if it persists.

**Comment by jonboh** [comment]
I've been running problematic test-suite continously since yesterday on both `main`and `0.19.0`. 
After ~1000 runs there are no errors so it seems to be fixed 👍
(In 0.17.1 I was getting a failure rate of ~2.5%)

**Comment by marcusbuffett** [comment]
Can confirm fixed on my end too. Running with 8 B200s and no crashes! Huge!

**Comment by laggui** [comment]
@jonboh @marcusbuffett thank you both for confirming!

Other users have reported an increased RAM usage with autodiff, and I noticed that not all graph nodes are being correctly dropped when no longer useful with the new graph management. So just a heads up that the RAM usage might continually increase in the current state 😬 

But we are working on a fix for that, and a patch 0.19.1 should come around soon.
